internal/states/encryption/instance.go

```diff
@@ -0,0 +1,50 @@
+package encryption
+
+import (
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+	"github.com/opentofu/opentofu/internal/states/encryption/flow"
+)
+
+// Instance obtains the instance of the encryption flow for the given configKey.
+//
+// configKey specifies a resource that accesses remote state. It must contain at least one ".".
+//
+// If the key is "terraform_remote_state.foo", the returned Flow is intended for
+//
+//	data "terraform_remote_state" "foo" {}
+//
+// For enumerated resources, the format is "terraform_remote_state.foo[17]" or
+// "terraform_remote_state.foo[key]" (no quotes around the for_each key).
+//
+// The first time a particular instance is requested, Instance may bail out due to invalid configuration.
+//
+// See also RemoteStateInstance, StatefileInstance, PlanfileInstance.
+func Instance(configKey string) flow.Flow {
```

Comment:
This function should maybe be called `NewInstance` to follow Go conventions?

Reply:
Once all TODO comments are implemented, it will only return a new instance if one wasn't found in the cache (for each configKey, the instance is a singleton so it can keep track of the configuration for this configKey).

NewOrCachedInstance? My train of thought was, it's "THE" instance for this configKey anyway.

Reply:
I'm generally not a fan of side effects. Caching in this way, without an injected cache layer, would make testing very hard since the behavior would depend on the caching. Generally, I prefer my caching to be an overlay that implements the same function as the underlying implementation, but then again, I tend tk write Go in a fairly OOP way which may not be the most appropriate.

Reply:
Not 100% sure, but in this case I think it's not too bad, because the only thing being cached is the configuration, and given an instance you can override all parts of the configuration by calling `EncryptionConfiguration(...)` and `DecryptionFallbackConfiguration(...)` with the relevant values of `source` on it before calling the code to be tested. In this way, tests can also avoid having to set environment variables, unless parsing them is the specific thing being tested, which is part of the unit tests of package `encryptionconfig` and `encryption`.

The tradeoff is that we can avoid passing around `Flow` instances between where they'll likely be configured ...
- parsing the `terraform {}` block, see `internal/configs/parser_config.go`
- remote_state data source configuration, see `internal/builtin/providers/tf`

... and where they'll likely be used
- `internal/states/remote/state.go`
- `internal/states/statefile/(read.go,version4.go)`
- `internal/plans/planfile/(reader.go,writer.go)`

We will see if this tradeoff is worth it when we implement the callsites (subtasks 2 and dependent in #1030). If it turns out to be doable to store the `Flow` instances externally without touching large bodies of existing code, we can remove the caching layer.

More reason to keep it isolated and small.

Reply:
I have renamed the Instance...(...) functions to "GetSingleton...(...)". I have also added a comment that explains the reason for having the singleton instance cache.

Reply:
I believe we sorted this out.